docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,226 @@
+
+# Categories of gadgets
+
+The category is automatically found with the following algorithm:
+1. There is a eBPF map with the prefix "print_":
+   1. The map type is RingBuf or PerfEventArray. Category: tracer.
+   2. The map type is Hash or Array. Category: topper.
+2. There is no eBPF map with the prefix "print_".
+   1. The is a eBPF program of type "iter". Category: snapshotter.
+
+## Tracer (stream of events)
+
+IG notices eBPF maps of type RingBuf or PerfEventArray with the prefix "print_". Then, IG checks the BTF type of values
+of the map. It uses this BTF information to parse the buffers from the ring buffers.
+
+## Snapshotters
+
+IG notices eBPF programs of type "iter/". Here we don't have maps with BTF information. IG runs the program and parses
+the output from BPF_SEQ_PRINTF() calls in the following format:
+```
+[event]
+fieldName1=fieldValue1\0
+fieldName2=fieldValue2\0
+```
+
+Note that some fields such as `comm` can contain a new line character ("\n"), so we can't use only that as a separator.
+
+Alternatively, the bpf module can contain a global variable with the same name as the iterator plus a suffix "_output".
+So the BTF information can be attached to that variable.
+```
+const volatile event_t myiterator_output = {};
+
+SEC("iter/task_file")
+int myiterator(struct bpf_iter__task_file *ctx)
+```
+
+Problem: if that variable is not used by any program, the cilium/ebpf loader will complain.
+
+## Toppers
+
+TODO
+
+# Reimplementing existing gadgets as containerized gadgets
+
+For each existing gadget, we list the additional feature needed that are not yet supported for containerized containers
+in Inspektor Gadget. For this exercise, we look at each gadget tracer.go and *.bpf.c file.
+
+| Gadget                 | Feature needed                                                                  | Difficulty |
+|------------------------|---------------------------------------------------------------------------------|------------|
+| advise network-policy  |                                                                                 |            |
+| advise seccomp-profile |                                                                                 |            |
+| audit seccomp          |                                                                                 |            |
+| profile block-io       |                                                                                 |            |
+| profile cpu            |                                                                                 |            |
+| profile tcprtt         |                                                                                 |            |
+| snapshot process       |                                                                                 |            |
+| snapshot socket        |                                                                                 |            |
```

Comment:
I think it's the most challenging one since its mechanism to run the iterator for each network namespace.

Reply:
I think it is simple to do:
* iter/bpf_map_elem: run in current namespace
* iter/task_file and iter/task: run in host pidns (since pidns are hierarchical)
* iter/tcp and iter/udp: run in all netns (since netns are independent)

Reply:
mmm, good point. We can define that approach and implement that logic inside Inspektor Gadget.